sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -319,149 +328,193 @@ public static Read read() {
-        "Validation of query \"%1$s\" failed. If the query depends on an earlier stage of the"
-        + " pipeline, This validation can be disabled using #withoutValidation.";
+  /////////////////////////////////////////////////////////////////////////////
 
-    /**
-     * Disable validation that the table exists or the query succeeds prior to pipeline submission.
-     * Basic validation (such as ensuring that a query or table is specified) still occurs.
-     */
-    public Read withoutValidation() {
-      return toBuilder().setValidate(false).build();
-    }
+  /**
+   * Implementation of {@link #readRecords()}.
+   */
+  @AutoValue
+  public abstract static class ReadGenericRecords<T>
+      extends ReadBase<T, ReadGenericRecords<T>, ReadGenericRecords.Builder<T>> {
 
-    /**
-     * Disable <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs">flattening of
-     * query results</a>.
-     *
-     * <p>Only valid when a query is used ({@link #fromQuery}). Setting this option when reading
-     * from a table will cause an error during validation.
-     */
-    public Read withoutResultFlattening() {
-      return toBuilder().setFlattenResults(false).build();
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T>
+        extends ReadBuilderBase<ReadGenericRecords<T>, Builder<T>> {
+      abstract Builder<T> setParseFnFactory(
+          SerializableFunction<TableSchema, SerializableFunction<GenericRecord, T>> parseFnFactory);
+      abstract Builder<T> setCoder(Coder<T> coder);
     }
 
-    /**
-     * Enables BigQuery's Standard SQL dialect when reading from a query.
-     *
-     * <p>Only valid when a query is used ({@link #fromQuery}). Setting this option when reading
-     * from a table will cause an error during validation.
-     */
-    public Read usingStandardSql() {
-      return toBuilder().setUseLegacySql(false).build();
+    public ReadGenericRecords<T> withParseFnFactory(
+        SerializableFunction<TableSchema, SerializableFunction<GenericRecord, T>> parseFnFactory) {
+      return toBuilder().setParseFnFactory(parseFnFactory).build();
     }
 
-    /**
-     * Use new template-compatible source implementation.
-     *
-     * <p>Use new template-compatible source implementation. This implementation is compatible with
-     * repeated template invocations. It does not support dynamic work rebalancing.
-     */
-    @Experimental(Experimental.Kind.SOURCE_SINK)
-    public Read withTemplateCompatibility() {
-      return toBuilder().setWithTemplateCompatibility(true).build();
+    public ReadGenericRecords<T> withCoder(Coder<T> coder) {
+      return toBuilder().setCoder(coder).build();
     }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  abstract static class ReadBuilderBase<ReadT, SelfT extends ReadBuilderBase<ReadT, SelfT>> {
+    abstract SelfT setJsonTableRef(ValueProvider<String> jsonTableRef);
```

Comment:
I'm impressed that AutoValue can handle this!

Reply:
lol it took me like an hour to coax AutoValue into getting this right.  Overriding `toBuilder()` seemed like the magic incantation to get it to compile.